### PR TITLE
Add fallback service for the `https` option

### LINKED ISSUE #32: Add more public IP checks for better resilience
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/public-ip/issues/32)

See https://github.com/sindresorhus/is-online/issues/64. One or two alternative services to be checked in parallel would probably be beneficial for better resilience against outages.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)
### Submitted pull Requests
- [#33 added backup service ipify in case primary service goes down](https://issuehunt.io/r/sindresorhus/public-ip/pull/33)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/public-ip/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by adityapatadia**
closes #32

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#32: Add more public IP checks for better resilience](https://issuehunt.io/repos/24034567/issues/32)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -76,9 +78,22 @@ const queryHttps = (version, options) => {
...
-			const {body} = await gotPromise;
+			let response;
+			try {
+				response = await gotPromise;
+			} catch (error) {
+				const gotBackupPromise = got(type[version].httpsBackupUrl, {
+					family: version === 'v6' ? 6 : 4,
+					retries: 0,
+					timeout: options.timeout
+				});
```
Comment (sindresorhus) [inline index.js]: The options are just duplicated and could be moved to a variable.

File: index.js [hunk 2]
```diff
@@ -17,15 +17,17 @@ const type = {
...
 	},
 	v6: {
 		dnsServer: '2620:0:ccc::2',
 		dnsQuestion: {
 			name: 'myip.opendns.com',
 			type: 'AAAA'
 		},
-		httpsUrl: 'https://ipv6.icanhazip.com/'
+		httpsUrl: 'https://ipv6.icanhazip.com/',
+		httpsBackupUrl: 'https://api6.ipify.org'
```
Comment (sindresorhus) [inline index.js]:
```suggestion
		httpsBackupUrl: 'https://api6.ipify.org/'
```

For consistency

File: browser.js [hunk 3]
```diff
@@ -10,22 +10,47 @@ const urls = {
...
+					doReject();
+				}
+
+				resolve(ip);
+			}, {once: true});
+
+			xhr.open('GET', backupUrls[version]);
+			xhr.timeout = options.timeout;
+			xhr.send();
+		};
```
Comment (sindresorhus) [inline browser.js]: Too much duplication. You could generalize the XHR stuff into a reusable async function that could be used for both the main and backup.

**Comment by sindresorhus** [comment]
You also need to update the readme about the backup.

File: browser.js [hunk 4]
```diff
@@ -10,22 +10,47 @@ const urls = {
 	v6: 'https://ipv6.icanhazip.com/'
 };
 
+const backupUrls = {
```
Comment (sindresorhus) [inline browser.js]: I think we should use the wording `fallback` instead of `backup` (everywhere).

Comment (fisker) [reply browser.js]: why not an array like

```js
const fallbackUrlsV4 = ['https://api.ipify.org']
const fallbackUrlsV6 = ['https://api6.ipify.org']
```

File: index.js (hunk 2 above)
Comment (adityapatadia) [reply index.js]: done

**Comment by adityapatadia** [comment]
Done. Please check.

File: browser.js [hunk 5]
```diff
@@ -10,31 +10,52 @@ const urls = {
...
 
-		xhr.open('GET', urls[version]);
+		xhr.open('GET', url);
 		xhr.timeout = options.timeout;
 		xhr.send();
 	});
+};
+
+const queryHttps = async (version, options) => {
+	const promise = new Promise((resolve, reject) => {
```
Comment (sindresorhus) [inline browser.js]: The `new Promise` wrapper here is moot. You can do it with just a normal async function.

File: index.js [hunk 6]
```diff
@@ -68,17 +70,28 @@ const queryHttps = (version, options) => {
 
 	const promise = (async () => {
 		try {
-			const gotPromise = got(type[version].httpsUrl, {
+			const requestOpts = {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
			const requestOptions = {
```

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/public-ip/pull/33#issuecomment-514109102 is not done.

File: browser.js [hunk 7]
```diff
@@ -10,37 +10,52 @@ const urls = {
...
-	};
+const queryHttps = async (version, options) => {
+	let ip;
+	try {
+		ip = await sendXhr(urls[version], options, version);
+	} catch (error) {
+		try {
+			ip = await sendXhr(fallbackUrls[version], options, version);
+		} catch (error) {
+			return Promise.reject(new Error('Couldn\'t find your IP'));
```
Comment (sindresorhus) [inline browser.js]:
```suggestion
			throw new Error('Couldn\'t find your IP');
```

File: browser.js [hunk 8]
```diff
@@ -10,37 +10,52 @@ const urls = {
...
+};
 
-	promise.cancel = () => {
-		xhr.abort();
-	};
+const queryHttps = async (version, options) => {
+	let ip;
+	try {
+		ip = await sendXhr(urls[version], options, version);
+	} catch (error) {
```
Comment (sindresorhus) [inline browser.js]:
```suggestion
	} catch (_) {
```

File: browser.js [hunk 9]
```diff
@@ -10,37 +10,52 @@ const urls = {
...
+		try {
+			ip = await sendXhr(fallbackUrls[version], options, version);
+		} catch (error) {
```
Comment (sindresorhus) [inline browser.js]:
```suggestion
		} catch (_) {
```

File: index.js [hunk 10]
```diff
@@ -68,17 +70,28 @@ const queryHttps = (version, options) => {
...
+
+			const gotPromise = got(type[version].httpsUrl, requestOptions);
 
 			cancel = gotPromise.cancel;
 
-			const {body} = await gotPromise;
+			let response;
+			try {
+				response = await gotPromise;
+			} catch (error) {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
			} catch (_) {
```

File: readme.md [hunk 11]
```diff
@@ -45,6 +45,8 @@ Default: `false`
 
 Use a HTTPS check using the [icanhazip.com](https://github.com/major/icanhaz) service instead of the DNS query. This check is much more secure and tamper-proof, but also a lot slower. **This option is only available in the Node.js version**.
 
+This library uses [ipify.org](https://www.ipify.org) as a fallback service for HTTPS check. In case `icanhazip.com` fails, this library will automatically query `ipify.org` service.
```
Comment (sindresorhus) [inline readme.md]: This could be written in a much shorter manner as part of the previous paragraph (line 46).

**Comment by adityapatadia** [comment]
I hope this can be merged now. I have done all requested changes. Thanks for updating readme for me.

**Comment by sindresorhus** [comment]
Yup. Looks good now. Thanks for your hard work :)

**Pull request merged**